File: 102. Binary Tree Level Order Traversal/102. Binary Tree Level Order Traversal.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 * BFS Queue
 * In each level, q.size()--> number of nodes
 */
class Solution {
    public List<List<Integer>> levelOrder(TreeNode root) {
        List<List<Integer>> ans = new ArrayList<>();
        if(root == null) return ans;
        
        Queue<TreeNode> q = new LinkedList<>();
        q.offer(root);
        
        while(! q.isEmpty()){
            int nums = q.size();
            List<Integer> list = new ArrayList<>();
            for(int i = 0; i < nums; i++){
                TreeNode curr = q.poll();
                list.add(curr.val);
                if(curr.left != null) q.offer(curr.left);
                if(curr.right != null) q.offer(curr.right);
            }
            ans.add(list);
        }
        return ans; 
    }
}

File: 102. Binary Tree Level Order Traversal/102. Binary Tree Level Order Traversal.js
/**
 * Definition for a binary tree node.
 * function TreeNode(val) {
 *     this.val = val;
 *     this.left = this.right = null;
 * }
 */
/**
 * @param {TreeNode} root
 * @return {number[][]}
 * Recursion（dfs)
 * Record the level of each node and directly add to the corresponding list
 */
var levelOrder = function(root) {
  let ans = [];
  if (root == null) return ans;
  traversal(root, 0, ans);
  return ans;
};

var traversal = function(node, level, ans) {
  if (ans.length <= level) {
    ans.push([]);
  }
  ans[level].push(node.val);

  if (node.left != null) traversal(node.left, level + 1, ans);
  if (node.right != null) traversal(node.right, level + 1, ans);
};
